### ISSUE TITLE
Update Customer data with REST API

### ISSUE THREAD

**Issue by judaschwartz** [magento/magento2#21237]
### Preconditions (*)

1. 2.3.0 (have not tested previous versions)

### Steps to reproduce (*)

1. When using the API to update a customer

     curl -X PUT "https://example.com/index.php/rest/all/V1/customers/8" -H "accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer TOKEN" -d "{ \"customer\": { \"middlename\": \"string\" }}"

### Expected result (*)

1. The customer should be updated for those fields just as when using the API to update a product  a PUT request like this 

    curl -X PUT "https://example.com/index.php/rest/all/V1/products/24-MB01" -H "accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer TOKEN" -d "{ \"product\": { \"visibility\": 3 }}"

updates the visibility for the product with that SKU (as one would expect) however when i try to make a similar update to a customer with a PUT request i get:
### Actual result (*)

1. I get a 400 error with `"message": "The customer email is missing. Enter and try again."` the API to update only works if i add all of the required data like this:

    curl -X PUT "https://example.com/index.php/rest/all/V1/customers/8" -H "accept: application/json" -H "Content-Type: application/json" -H "Authorization: Bearer TOKEN" -d "{ \"customer\": { \"id\": \"8\", \"firstname\": \"Jon\", \"lastname\": \"Smith\", \"email\": \"Jjsmith@gmail.com\", \"middlename\": \"jac\", \"store_id\": 1, \"website_id\": 1 }}"

if I leave any of these values out i get an error with the exception of `id` which if left out creates a new record or throws an error if the email already exists (despite the `id` being specified at the end of the URL) also when `id` is passed in with the data that IDs record is updated even if it is not the `id` at the end of the URL

This does not seem to be the way the API should operate any help or explanation of whats going on would be appreciated.

**Comment by magento-engcom-team** [comment]
Hi @judaschwartz. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@judaschwartz do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by judaschwartz** [comment]
`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

**Comment by magento-engcom-team** [comment]
Hi @judaschwartz. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @judaschwartz, here is your [Magento instance](https://i-21237-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-21237-2-3-develop.instances.magento-community.engineering/admin](https://i-21237-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by judaschwartz** [comment]
* Reproduced on 2.2.x

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-98296`, `MAGETWO-98297` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by judaschwartz** [comment]
I created a PR to fix this
https://github.com/magento/magento2/pull/21236

**Comment by mmihalev** [comment]
I have exactly the same issue. Please fix.

**Comment by judaschwartz** [comment]
> I have exactly the same issue. Please fix.

I made a PR to fix this over a year ago, someone put the PR on hold maybe if you comment on that PR #21236 maybe that will get it unstuck and merged

**Comment by mmihalev** [comment]
> > I have exactly the same issue. Please fix.
> 
> I made a PR to fix this over a year ago, someone put the PR on hold maybe if you comment on that PR #21236 that will get it unstuck and merged

... unbelievable ...

**Comment by gotanasro3** [comment]
please fix it

**Comment by slavvka** [comment]
Hi @judaschwartz. Thank you for your report. 
The issue has been fixed in magento/magento2#28332 by @engcom-Charlie in 2.4-develop branch
Related commit(s):
 - [4afc05fe11dfa954f27a4585d0cd33ce55b5c85a](https://github.com/magento/magento2/commit/4afc05fe11dfa954f27a4585d0cd33ce55b5c85a)
 - [d6d15d792f2cb4237dacc20a9f23bb1c952da503](https://github.com/magento/magento2/commit/d6d15d792f2cb4237dacc20a9f23bb1c952da503)
 - [cdbf98876595888ba9c883b77ca09e15065f41be](https://github.com/magento/magento2/commit/cdbf98876595888ba9c883b77ca09e15065f41be)
 - [fe62842511c6da650f62020f17425730c03dd64f](https://github.com/magento/magento2/commit/fe62842511c6da650f62020f17425730c03dd64f)

The fix will be available with the upcoming 2.4.1 release.
